### PR TITLE
schemahcl: fixed invalid identity token

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

**Review by rotemtam — COMMENTED** [review]
Well done! Can you add an integration test using testscript to verify this works e2e on postgres/mysql?

**Comment by giautm** [comment]
> Well done! Can you add an integration test using testscript to verify this works e2e on postgres/mysql?

yes, please me try. :D

File: schemahcl/hcl.go [hunk 1]
```diff
@@ -541,6 +538,49 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
 	return nil, false
 }
 
+func hclRefTokens(ref string) hclwrite.Tokens {
+	s := strings.Split(strings.ReplaceAll(ref, "$", ""), ".")
+	t := make([]*hclwrite.Token, 0, len(s))
+	for i, s := range s {
+		if i == 0 {
```
Comment (rotemtam) [inline schemahcl/hcl.go:545]: Maybe add comments explaining the cases

File: schemahcl/hcl.go [hunk 2]
```diff
@@ -541,6 +538,49 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
...
+		} else if hclsyntax.ValidIdentifier(s) {
+			t = append(t, hclTraverseAttr(s)...)
+		} else {
+			t = append(t, hclTraverseIndex(cty.StringVal(s))...)
+		}
+	}
+	return t
+}
+
+func hclTraverseAttr(name string) hclwrite.Tokens {
```
Comment (rotemtam) [inline schemahcl/hcl.go:556]: Why "traverse"?

Comment (giautm) [reply schemahcl/hcl.go:556]: I used the term from `NewExpressionAbsTraversal`  https://github.com/hashicorp/hcl/blob/v2.13.0/hclwrite/ast_expression.go#L87-L119

File: schemahcl/hcl.go (hunk 1 above)
Comment (giautm) [reply schemahcl/hcl.go:545]: Root element in NewExpressionAbsTraversal

File: internal/integration/testdata/mysql/column-add-drop-2.txt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+# Compare database with 2.sql.
+cmpshow users 2.sql
+
+# Apply schema "1.hcl" should migrate database to previous state.
+apply 1.hcl
+cmpshow users 1.sql
+
+# Drop table.
+apply 0.hcl
+! exist users
```
Comment (a8m) [inline internal/integration/testdata/mysql/column-add-drop-2.txt:26]: Add another assertion using `cmphcl` to ensure HCL writing is correct. Also, no need for another test, you can extend the current one.

File: schemahcl/hcl.go [hunk 4]
```diff
@@ -541,6 +538,49 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
...
+	s := strings.Split(strings.ReplaceAll(ref, "$", ""), ".")
+	t := make([]*hclwrite.Token, 0, len(s))
+	for i, s := range s {
+		if i == 0 {
+			t = append(t, hclRawTokens(s)...)
+		} else if hclsyntax.ValidIdentifier(s) {
+			t = append(t, hclTraverseAttr(s)...)
+		} else {
+			t = append(t, hclTraverseIndex(cty.StringVal(s))...)
+		}
```
Comment (a8m) [inline schemahcl/hcl.go:551]: Use `switch` and inline `hclTraverseAttr` as it's used only here.

File: schemahcl/hcl.go [hunk 5]
```diff
@@ -541,6 +538,49 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
...
+			Bytes: []byte{'.'},
+		},
+		&hclwrite.Token{
+			Type:  hclsyntax.TokenIdent,
+			Bytes: []byte(name),
+		},
+	}
+}
+
+func hclTraverseIndex(v cty.Value) hclwrite.Tokens {
```
Comment (a8m) [inline schemahcl/hcl.go:569]: ditto.

File: schemahcl/hcl.go [hunk 6]
```diff
@@ -541,6 +538,49 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
 	return nil, false
 }
 
+func hclRefTokens(ref string) hclwrite.Tokens {
+	s := strings.Split(strings.ReplaceAll(ref, "$", ""), ".")
```
Comment (a8m) [inline schemahcl/hcl.go:542]: `"c $"` is also a valid (quoted) identifier. Let's handle it as well?

**Review by a8m — APPROVED** [review]
🚀

**Review by a8m — CHANGES_REQUESTED** [review]
Accidentally approved 😃

File: internal/integration/testdata/mysql/column-add-drop-2.txt (hunk 3 above)
Comment (a8m) [reply internal/integration/testdata/mysql/column-add-drop-2.txt:26]: You can extend this one: https://github.com/ariga/atlas/blob/master/internal/integration/testdata/mysql/column-generated-inspect.txt#L5

File: schemahcl/hcl.go [hunk 7]
```diff
@@ -541,6 +538,39 @@ func findAttr(attrs []*Attr, k string) (*Attr, bool) {
 	return nil, false
 }
 
+func hclRefTokens(ref string) hclwrite.Tokens {
+	t := []*hclwrite.Token{}
```
Comment (a8m) [inline schemahcl/hcl.go:542]:
```suggestion
	var t []*hclwrite.Token
```

**Review by a8m — APPROVED** [review]
:shipit: ship it!

**Pull request merged**